### PR TITLE
[DB Import] Add presubmit check

### PR DESCRIPTION
**Pull request by beckerhe**
This PR is adding a few related things at the same time (Apologies for that.):

1. It introduces a new CLI command `config` which adds to utility functions for handling config files: `cli.py config dump` dumps the entire config after YAML processing and `cli.py config list_pipelines` returns a newline-separated list with all the defined pipelines.
2. It is adding a Docker image definition for running presubmit checks in a Docker container. The purpose of the Docker container is mainly to build the bigquery-emulator. (It's a golang-tool and needs to be built from source)
3. It is adding a GitHub Actions workflow that first builds the Docker image and then runs the unit and integration tests.
4. It is adding a dummy pipeline config which is needed to see whether the integration tests work.

### CODE REVIEW

File: .github/workflows/db_import.yml
Comment (pzread) [inline .github/workflows/db_import.yml:43]: Ah... This is super cool. Thanks for researching this method.

I'm slightly worried about two points:

1. For reproducibility, users might want to find the same `image@hash` to pull from a registry as CI, is it still possible (or easy) with Github action cache?
2. It looks like the max total cache size of a repo is 10GB (https://docs.github.com/en/actions/using-workflows/caching-dependencies-to-speed-up-workflows#usage-limits-and-eviction-policy), I'm wondering if we max out soon and see cache misses. (I can see ~700MB is used from this PR).

Regarding the Github cache, although it is safe to use now, I have a plan to change the security model of our CI (in order to improve efficiency) which might make it unsafe to use (cache poisoning from untrusted runners). I'm happy to explain it more but it's another reason we might not want to use it right now.

Just FYI (you might already know), in this repo we have added the IREE-style docker image managing tool: https://github.com/openxla/openxla-benchmark/blob/main/devtools/docker/manage_images.py. You add docker images to build in `devtools/docker/image_deps.json` and `manage_images.py` will build and update the references in repo.

File: .github/workflows/db_import.yml
Comment (pzread) [inline .github/workflows/db_import.yml:58]: We usually mount the repo dir to `/work` in the docker. This provides the same setup as the host (runner). It will avoid some issues when running with generated artifacts having hard-coded `/work` paths.

https://github.com/openxla/openxla-benchmark/blob/b24f192752812e3dd4211ed4347d6d64e6774950/.github/workflows/run_comparative_benchmark.yml#L82-L86

File: devtools/db_import/Dockerfile.presubmit
Comment (pzread) [inline devtools/db_import/Dockerfile.presubmit:1]: It looks like our naming convention for dockerfile is `*.Dockerfile` (copied from IREE)

https://github.com/openxla/openxla-benchmark/blob/main/devtools/docker/dockerfiles/base.Dockerfile

And should we put all dockerfiles under https://github.com/openxla/openxla-benchmark/blob/main/devtools/docker/dockerfiles/?

File: .github/workflows/db_import.yml
Comment (pzread) [inline .github/workflows/db_import.yml:6]: Maybe we also want to run when PR is targeting other branches? (useful when we try to stack a series of PRs on branches).

File: .github/workflows/db_import.yml
Comment (pzread) [inline .github/workflows/db_import.yml:13]: Maybe also add `workflow_dispatch` for manual trigger?

**Review by pzread — COMMENTED** [review]
Sorry for the delay. Mostly look good. I have some comments on handling docker images.

File: .github/workflows/db_import.yml
Comment (beckerhe) [reply .github/workflows/db_import.yml:43]: Regarding 1. Yes that's possible and easy to do, just not implemented yet. The usual approach would be to also push to a registry when running from main (no PR), so that users can get the container from the registry. I haven't implemented that yet, because it needs authentication.

Regarding 2. Yes, total cache size is a concern, especially when we use that technique for more than this one docker image. But still, as long as we don't make frequent changes to this docker image, the total cache usage shouldn't go beyond these 700MiB. Once this PR gets merged, this PR's cache gets evicted and the postsubmit workflow will create a cache on the main branch. This cache will then be reused for all following PRs. As long as there is no change to the Docker container, no new cache artifacts are created. The cache usage can be further reduced by not pushing to the cache from the main branch, but instead only push to the registry from the main branch. Each PR will then start with an empty cache and download the cached layers from the registry. If there were changes to the Docker image definition, the PR's cache would preserve the image between CI runs. That way the cache only needs to do its thing for the lifetime of a PR - and only for PRs that actually make changes to a cached Docker image.

Regarding cache poisoning: I'm not sure if I understand what your plans are, but I think cache poisining is not a big issue with GitHub caches, because the cache if not shared. Each PR and branch have their own cache, meaning one PR can't affect the cached artifacts of another PR. But when a new PR gets created its cache gets prefilled with the main branch's cache, so that you don't start with an empty cache.

Regarding `manage_images.py`. Yes, I saw that tool and I'm happy to switch to that if you prefer that. I wanted though to give `docker buildx` with its enhanced caching a try. It makes changes to the Docker container fully self-served. No need for a maintainer to push a new image or update a digest. Everything gets tested on the CI and it's possible to make changes to a Docker image and some other part of the repo at the same time in the same commit.

File: .github/workflows/db_import.yml
Comment (beckerhe) [reply .github/workflows/db_import.yml]: Changed. Thanks.

File: devtools/db_import/Dockerfile.presubmit
Comment (beckerhe) [reply devtools/db_import/Dockerfile.presubmit:1]: Done.

File: .github/workflows/db_import.yml
Comment (beckerhe) [reply .github/workflows/db_import.yml]: No objections. I removed the branch filter.

File: .github/workflows/db_import.yml
Comment (beckerhe) [reply .github/workflows/db_import.yml:11]: Done.

File: .github/workflows/db_import.yml
Comment (beckerhe) [reply .github/workflows/db_import.yml:20]: Makes sense. Changed.

File: .github/workflows/db_import.yml
Comment (beckerhe) [reply .github/workflows/db_import.yml:49]: I just realized I might be wrong on how cache keying actually works since Docker Buildx is directly talking to the caching API. Let me double check that.

Comment (pzread) [reply .github/workflows/db_import.yml:49]: > The usual approach would be to also push to a registry when running from main (no PR), so that users can get the container from the registry. I haven't implemented that yet, because it needs authentication.

How do users find the digest to pull the image? Currently that is clear because we embed the digests in the workflow files and they are tracked in git history. If we don't track digests in commits (and just have the tag `presubmit:current` for the latest one), it might be a problem to reproduce the exact environment at a previous commit, as users don't know which digest was used in that commit.

> Yes, total cache size is a concern, especially when we use that technique for more than this one docker image. But still, as long as we don't make frequent changes to this docker image, the total cache usage shouldn't go beyond these 700MiB.

I prefer to have one method to manage all docker images in a repo, which means if we move to use `docker/build-push-action`, we might want to upload all docker images to the cache. This might quickly saturate the cache limit. For example the image we are using in benchmark `gcr.io/iree-oss/openxla-benchmark/openxla-benchmark-cuda11.8-cudnn8.9` takes 18.2GB (I guess due to CUDA)

For the security concern, right now our CI security model assumes that all postsubmit jobs are running on trusted runners. That creates some problems, for example, the testing/benchmark runners can't be permanent and shared between pre and postsubmit (for efficiency), we can't accept external partners to contribute testing/benchmark runners. Theoretically, we should only need to ensure the building and release jobs are running on trustable runners (so our artifacts are trustable), the testing and benchmark jobs can be running on untrustable ones (since they only upload JSON files, which don't pose much risk).

This change means we will mix trustable and untrustable jobs in postsubmit workflows. So if Github action cache can't prevent untrustable jobs in the same workflow to poison caches, it can be a blocker for us to move to this less-restricted security model. And AFAIK that's the case (https://scribesecurity.com/blog/github-cache-poisoning).

The security concern shouldn't be a deal breaker here since the new security model is still in design. But I'd like to raise it here for the awareness.

**Review by pzread — APPROVED** [review]
LGTM as there is no immediate issue and I don't want to block this for too long.

Regarding the docker images, I'm happy to explore new method to manage them (and it really improves the development workflow). But I think P0 is to provide users the digests to pull docker images from gcr. So if we don't solve it in this PR, makes sure we have an assigned issue to track the follow-up.

File: .github/workflows/db_import.yml
Comment (beckerhe) [reply .github/workflows/db_import.yml:49]: > > The usual approach would be to also push to a registry when running from main (no PR), so that users can get the container from the registry. I haven't implemented that yet, because it needs authentication.
> 
> How do users find the digest to pull the image? Currently that is clear because we embed the digests in the workflow files and they are tracked in git history. If we don't track digests in commits (and just have the tag `presubmit:current` for the latest one), it might be a problem to reproduce the exact environment at a previous commit, as users don't know which digest was used in that commit.

Users would rebuild the docker container from source using `docker buildx build` (probably wrapped in a script). And `docker buildx` would determine hashes for each layer and then pull those layers from the registry (or actually build them from source if they are not cached.)

> 
> > Yes, total cache size is a concern, especially when we use that technique for more than this one docker image. But still, as long as we don't make frequent changes to this docker image, the total cache usage shouldn't go beyond these 700MiB.
> 
> I prefer to have one method to manage all docker images in a repo, which means if we move to use `docker/build-push-action`, we might want to upload all docker images to the cache. This might quickly saturate the cache limit. For example the image we are using in benchmark `gcr.io/iree-oss/openxla-benchmark/openxla-benchmark-cuda11.8-cudnn8.9` takes 18.2GB (I guess due to CUDA)

Using a registry for caching (https://docs.docker.com/build/cache/backends/registry/) instead of GitHub Actions could solve that. I'm not sure though how garbage collection works in that case.

What would it entail for me to switch to the digests tool? I guess you have to build the image and upload it and send me a digest which I can put into this PR, right?

> For the security concern, right now our CI security model assumes that all postsubmit jobs are running on trusted runners. That creates some problems, for example, the testing/benchmark runners can't be permanent and shared between pre and postsubmit (for efficiency), we can't accept external partners to contribute testing/benchmark runners. Theoretically, we should only need to ensure the building and release jobs are running on trustable runners (so our artifacts are trustable), the testing and benchmark jobs can be running on untrustable ones (since they only upload JSON files, which don't pose much risk).
> 
> This change means we will mix trustable and untrustable jobs in postsubmit workflows. So if Github action cache can't prevent untrustable jobs in the same workflow to poison caches, it can be a blocker for us to move to this less-restricted security model. And AFAIK that's the case (https://scribesecurity.com/blog/github-cache-poisoning).
> 
> The security concern shouldn't be a deal breaker here since the new security model is still in design. But I'd like to raise it here for the awareness.

Not sure if there is a way to restrict write access to the cache based on runner or workflow. If so that would solve it, I guess?. Another option is to use registry based caching and GCE machine authentication. In that case only our trusted runners would get write access to the registry/cache.

Comment (pzread) [reply .github/workflows/db_import.yml:49]: Agree, I feel like most of the problems are related to Github action cache and `docker/build-push-action` is actually working well. Many problems can probably be solved by switching registry to our own gcr with GCE machine authentication.

Comment (beckerhe) [reply .github/workflows/db_import.yml:49]: Considering our requirements I see 2 ways forward:

1. Rewrite it to use the proven method (`prod_digests.txt`)
2. Move to registry based caching and move the cuda container to use the same (probably in a separate PR)

**Comment by beckerhe** [comment]
> But I think P0 is to provide users the digests to pull docker images from gcr. So if we don't solve it in this PR, makes sure we have an assigned issue to track the follow-up.

If it has such a high priority I think we should try to solve it immediately or at least have a plan before merging this.

If we wanted to use the "have the CI build the containers" approach that I'm proposing here I see the following options. (Of course it's also possible to just use the proven approach with `prod_digests.txt`):

1. Have also the user build their own docker images on demand. This would mean we have a script which calls `docker buildx build` instead of `docker pull`. Docker would then "build" the image but since we will have all the caching layers in the registry this will end up being the same image as if it was pulled. The only difference is if a layer is not in the cache - a `docker pull` would fail, but a `docker buildx build` would just build the layer. To me this is the preferred approach because the user will never have an outdated or non-matching docker image. `docker buildx build` will always fetch or build the proper version.
2. If we wanted to have tags the user can `docker pull` from, then one option would be to use commit hashes as image tags. In that case we would have a script that determines the commit hash from the git repo and looks up the corresponding docker image. This of course doesn't work if HEAD is not a commit on main, but we could fall back to the latest common ancestor commit between HEAD and origin/main.
3. Or we determine the latest commit that affected the docker image and use that. This has the disadvantage that now the script determining the tag to pull needs to know what could have affected the image - which is really not its job. But I see the biggest downside that the user needs to know when they should rebuild the Docker image because local non-committed changes won't affect the logic. 

So what I'm basically asking is if solution 1 would be an option or if we need prebuilt images. :sweat_smile: 
If we do then option 2 or 3 might be valid, but I would rather lean towards using `prod_digests.txt` for now and
maybe come back to the BuildKit based approach later.

WDYT?

**Comment by pzread** [comment]
> > But I think P0 is to provide users the digests to pull docker images from gcr. So if we don't solve it in this PR, makes sure we have an assigned issue to track the follow-up.
> 
> If it has such a high priority I think we should try to solve it immediately or at least have a plan before merging this.
> 
> If we wanted to use the "have the CI build the containers" approach that I'm proposing here I see the following options. (Of course it's also possible to just use the proven approach with `prod_digests.txt`):
> 
> 1. Have also the user build their own docker images on demand. This would mean we have a script which calls `docker buildx build` instead of `docker pull`. Docker would then "build" the image but since we will have all the caching layers in the registry this will end up being the same image as if it was pulled. The only difference is if a layer is not in the cache - a `docker pull` would fail, but a `docker buildx build` would just build the layer. To me this is the preferred approach because the user will never have an outdated or non-matching docker image. `docker buildx build` will always fetch or build the proper version.
> 2. If we wanted to have tags the user can `docker pull` from, then one option would be to use commit hashes as image tags. In that case we would have a script that determines the commit hash from the git repo and looks up the corresponding docker image. This of course doesn't work if HEAD is not a commit on main, but we could fall back to the latest common ancestor commit between HEAD and origin/main.
> 3. Or we determine the latest commit that affected the docker image and use that. This has the disadvantage that now the script determining the tag to pull needs to know what could have affected the image - which is really not its job. But I see the biggest downside that the user needs to know when they should rebuild the Docker image because local non-committed changes won't affect the logic.
> 
> So what I'm basically asking is if solution 1 would be an option or if we need prebuilt images. sweat_smile If we do then option 2 or 3 might be valid, but I would rather lean towards using `prod_digests.txt` for now and maybe come back to the BuildKit based approach later.
> 
> WDYT?



> > But I think P0 is to provide users the digests to pull docker images from gcr. So if we don't solve it in this PR, makes sure we have an assigned issue to track the follow-up.
> 
> If it has such a high priority I think we should try to solve it immediately or at least have a plan before merging this.
> 
> If we wanted to use the "have the CI build the containers" approach that I'm proposing here I see the following options. (Of course it's also possible to just use the proven approach with `prod_digests.txt`):
> 
> 1. Have also the user build their own docker images on demand. This would mean we have a script which calls `docker buildx build` instead of `docker pull`. Docker would then "build" the image but since we will have all the caching layers in the registry this will end up being the same image as if it was pulled. The only difference is if a layer is not in the cache - a `docker pull` would fail, but a `docker buildx build` would just build the layer. To me this is the preferred approach because the user will never have an outdated or non-matching docker image. `docker buildx build` will always fetch or build the proper version.
> 2. If we wanted to have tags the user can `docker pull` from, then one option would be to use commit hashes as image tags. In that case we would have a script that determines the commit hash from the git repo and looks up the corresponding docker image. This of course doesn't work if HEAD is not a commit on main, but we could fall back to the latest common ancestor commit between HEAD and origin/main.
> 3. Or we determine the latest commit that affected the docker image and use that. This has the disadvantage that now the script determining the tag to pull needs to know what could have affected the image - which is really not its job. But I see the biggest downside that the user needs to know when they should rebuild the Docker image because local non-committed changes won't affect the logic.
> 
> So what I'm basically asking is if solution 1 would be an option or if we need prebuilt images. sweat_smile If we do then option 2 or 3 might be valid, but I would rather lean towards using `prod_digests.txt` for now and maybe come back to the BuildKit based approach later.
> 
> WDYT?

I'm actually convinced solution 1 is a good idea, but there are a few things need to be done:

1. Create and use our own gcr as cache, as github action cache is limited
2. A wrapper script to help users run locally.
3. Figure out how to build dependency between dockerfiles (e.g. https://github.com/openxla/openxla-benchmark/blob/main/devtools/docker/dockerfiles/cuda11.8-cudnn8.9.Dockerfile#L9) if there is no digest reference from the gcr. It's anther major reason we introduced `manage_images.py`. Maybe https://www.docker.com/blog/dockerfiles-now-support-multiple-build-contexts/#:~:text=Create%20Build%20Pipelines%20by%20Linking%20bake%20Targets can be a solution.

So maybe let's use `prod_digests.txt` and we need more time to prepare for solution 1. (Also if you have time, could you also create an issue to switch to solution 1?)

File: devtools/docker/dockerfiles/db_import.Dockerfile
Comment (beckerhe) [reply devtools/docker/dockerfiles/db_import.Dockerfile]: Done

File: .github/workflows/db_import.yml
Comment (beckerhe) [reply .github/workflows/db_import.yml]: Moved to prod_digests.txt. Thank you for the discussion. :-)

**Review by pzread — APPROVED** [review]

**Comment by pzread** [comment]
LGTM, Thanks!

**Pull request merged**
